src/L2VoteAggregator.sol

```diff
@@ -0,0 +1,145 @@
+
+import {L2GovernorMetadata} from "src/L2GovernorMetadata.sol";
+import {IL1Block} from "src/interfaces/IL1Block.sol";
+
+/// @notice A contract to collect votes on L2 to be bridged to L1.
+contract L2VoteAggregator {
+  /// @notice The number of blocks before L2 voting closes. We close voting 1200 blocks before the
+  /// end of the proposal to cast the vote. We will allow another 1200 blocks to allow for the final
+  /// vote total to be sent to the L1.
+  uint32 public constant CAST_VOTE_WINDOW = 2400;
+
+  /// @notice The Wormhole contract to bridge messages to L1.
+  IWormhole immutable CORE_BRIDGE;
+
+  /// @notice A unique number used to send messages.
+  uint32 public nonce;
+
+  /// @notice The token used to vote on proposals provided by the `GovernorMetadata`.
+  ERC20Votes immutable VOTING_TOKEN;
+
+  /// @notice The `GovernorMetadata` contract that provides proposal information.
+  L2GovernorMetadata immutable GOVERNOR_METADATA;
+
+  /// @notice The contract that handles fetch the L1 block on the L2.
+  IL1Block immutable L1_BLOCK;
+
+  /// @dev Thrown when an address has no voting weight on a proposal.
+  error NoWeight();
+
+  /// @dev Thrown when an address has already voted.
+  error AlreadyVoted();
+
+  /// @dev Thrown when an invalid vote is cast.
+  error InvalidVoteType();
+
+  /// @dev Thrown when proposal is inactive.
+  error ProposalInactive();
+
+  /// @dev The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @dev Data structure to store vote preferences expressed by depositors.
+  // TODO: Does it matter if we use a uint128 vs a uint256?
+  struct ProposalVote {
+    uint128 against;
+    uint128 inFavor;
+    uint128 abstain;
+  }
+
+  /// @notice A mapping of proposal to a mapping of voter address to boolean indicating whether a
+  /// voter has voted or not.
+  mapping(uint256 => mapping(address => bool)) private _proposalVotersHasVoted;
+
+  /// @notice A mapping of proposal id to proposal vote totals.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @param _votingToken The token used to vote on proposals.
+  /// @param _core The Wormhole contract to bridge messages to L1.
+  /// @param _governorMetadata The `GovernorMetadata` contract that provides proposal information.
+  /// @param l1BlockAddress The address of the L1Block contract.
+  constructor(
+    address _votingToken,
+    address _core,
+    address _governorMetadata,
+    address l1BlockAddress
+  ) {
+    VOTING_TOKEN = ERC20Votes(_votingToken);
+    CORE_BRIDGE = IWormhole(_core);
+    GOVERNOR_METADATA = L2GovernorMetadata(_governorMetadata);
+    L1_BLOCK = IL1Block(l1BlockAddress);
+  }
+
+  /// @notice Where a user can express their vote based on their L2 token voting power.
+  /// @param proposalId The id of the proposal to vote on.
+  /// @param support The type of vote to cast.
+  function expressVote(uint256 proposalId, uint8 support) external {
```

Comment:
We should use the same `castVote` signature as the Governor for this method. This will allow us to add the contract to frontends like Tally out of the box.

Reply:
https://github.com/ScopeLift/l2-flexible-voting/pull/9/commits/c372fed9112d5e83381fff50512aacc13f257213

Reply:
Is the expectation we support all of these methods on the `L2Aggregator` contract https://docs.tally.xyz/user-guides/tally-contract-compatibility/openzeppelin-governor#function-signatures

Reply:
That's a good question. I think we should consider it, but it's not critical for the moment. Why don't we pull it into a separate issue then we can resolve this convo.

Reply:
https://github.com/ScopeLift/l2-flexible-voting/issues/19